### PR TITLE
Optimize requires for non-omnibus installs

### PR DESCRIPTION
**Pull request by tas50**
require is quite slow in Ruby and doing requires for things you've already required is also slow. We've used this simple hack in Chef to speed up our requires. In the omnibus installs we patch how rubygems works to make this somewhat pointless, but this will help non-omnibus installs of InSpec.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/bundles/inspec-supermarket/api.rb
Comment (james-stocks) [inline lib/bundles/inspec-supermarket/api.rb:3]: C compiler directive vibes :)

**Comment by james-stocks** [comment]
This looks good to me, but might be wisest to merge after https://github.com/inspec/inspec/pull/5204

**Comment by Schwad** [comment]
Thank you for this @tas50 ! 🙏 

Great point @james-stocks . This would be a fantastic candidate to try out after we get 

* #5204 (all the tests)
* And an answer on releng's https://github.com/chef/release-engineering/issues/1236 to resolve https://github.com/inspec/inspec/issues/5003 

-- so we could have documented performance/improvement on this issue. 

For those unsure about the extra time lost doing unnecessary requires, here's an example you can fire off today in your terminal:

```
require 'benchmark/ips'


Benchmark.ips do |x|

  x.config(time: 10, warmup: 4)

  x.report("launchy1") do
    require "launchy"
  end

  x.report("launchy10") do
    10.times do
      require "launchy"
    end
  end

  x.report("launchy100") do
    100.times do
      require "launchy"
    end
  end

  x.report("launchy1000") do
    1000.times do
      require "launchy"
    end
  end

  x.compare!
end

Calculating -------------------------------------
            launchy1      1.361M (± 1.9%) i/s -     13.698M in  10.071021s
           launchy10    131.324k (± 2.3%) i/s -      1.319M in  10.050078s
          launchy100     13.034k (± 2.3%) i/s -    130.391k in  10.010020s
         launchy1000      1.303k (± 3.3%) i/s -     13.029k in  10.012206s

Comparison:
            launchy1:  1360698.2 i/s
           launchy10:   131323.6 i/s - 10.36x  slower
          launchy100:    13033.6 i/s - 104.40x  slower
         launchy1000:     1302.8 i/s - 1044.44x  slower
```

And a comparison with @tas50 's solution here:

```
require 'benchmark/ips'


Benchmark.ips do |x|

  x.config(time: 10, warmup: 4)

  x.report("launchy") do
    10.times do
      require "launchy"
    end
  end

  x.report("launchy unless defined") do
    10.times do
      require "launchy" unless defined?(Launchy)
    end
  end

  x.compare!
end

Calculating -------------------------------------
             launchy    133.487k (± 2.2%) i/s -      1.341M in  10.051712s
launchy unless defined
                          1.871M (± 2.4%) i/s -     18.787M in  10.045053s

Comparison:
launchy unless defined:  1871374.2 i/s
             launchy:   133487.3 i/s - 14.02x  slower
```

Traditionally I would've expected in a gem to utilize as many of our requires as possible in https://github.com/inspec/inspec/blob/master/lib/inspec.rb but with a gem this size I'm sure we get to a point where the requiring often happens on a file-level.

**Comment by Schwad** [comment]
#5204 will likely be getting merged in the near future, which moves this PR up. I wonder if this would also work using ruby's `autoload` or if that would make any difference. 🤔

**Comment by tas50** [comment]
Some of these we'd want to later convert to autoload, but some you just can't really autoload so these are still helpful.

**Comment by lamont-granquist** [comment]
@Schwad where autoload can be used it is even better than this.  That will prevent the gems from being even loaded once if they're never used, which really can't be beat.

**Comment by Schwad** [comment]
@lamont-granquist cool. I may fire that up as a test PR to see if that approach burns everything down. If not, it would be neat to run a quick manual benchmark of those two PRs vs. master on a compliance-profile-level run. (&& also a simple small run with one control)

**Comment by lamont-granquist** [comment]
Autoload isn't perfectly straightforwards text editing like this is.  You'll be better off shipping this as an incremental fix, then working on autoload.  We've autoloaded chef/chef and it just takes a week of beating on it or so working around the quirks of autoload.

**Comment by lamont-granquist** [comment]
(and you'll still wind up needing to preserve some of these)

**Comment by Schwad** [comment]
@lamont-granquist I do see what you mean. That test PR has required a bit of beating around but I might keep rolling with it.  Tentatively in my eyes this is "approved" pending 5204. Thanks all for the help and feedback

**Review by james-stocks — APPROVED** [review]
I believe the one risk that we're missing tests for (either currently or after #5204 merges) is that this repo breaks into 4 gems, and `require`s need to work after that split happens. The changes in this PR don't seem to overlap with that risk though, so this looks good to me!

**Pull request closed without merging**
